### ISSUE TITLE
Cast happens after copy in mixed precision training on IPU

### ISSUE THREAD

**Issue by HMellor** [Lightning-AI/lightning#13828]
## 🐛 Bug

During mixed precision training on IPU, the cast happens after the copy. When working in half precision on IPU it is preferred to cast to cast to half precision on the host.

### To Reproduce

<details>
<summary>
repro.py
</summary>

```python
import os
import torch
import argparse
import poptorch
import pytorch_lightning as pl
import torch.nn.functional as F
from torchvision import datasets, models
from torchvision.transforms import ToTensor
from pytorch_lightning.strategies.ipu import IPUStrategy


class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = models.resnet18(pretrained=False)

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.model(x)
        loss = F.cross_entropy(y_hat, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)


def main():
    batch_size = 32
    # data
    mnist = datasets.FakeData(
        size=100 * batch_size,
        image_size=(3, 28, 28),
        num_classes=10,
        transform=ToTensor(),
    )
    opts = poptorch.Options()
    train_loader = poptorch.DataLoader(
        options=opts,
        dataset=mnist,
        batch_size=batch_size,
        drop_last=True,
        mode=poptorch.DataLoaderMode.Async,
    )
    # model
    model = Model()
    # training
    trainer = pl.Trainer(
        accelerator="ipu",
        precision=args.precision,
        max_epochs=1,
        strategy=IPUStrategy(training_opts=opts),
    )
    trainer.fit(model, train_loader)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--precision", choices=[16, 32], default=16, type=int)
    args = parser.parse_args()
    os.environ["POPLAR_ENGINE_OPTIONS"] =f'{{"autoReport.all":"true","autoReport.directory":"profile_fp{args.precision}"}}'
    main()
```

</details>

```bash
python repro.py --precision 16
python repro.py --precision 32
```

Look at the profiles in PopVision Graph Analyser and see that the copy is FP32 followed by a cast to FP16 on the IPU.

### Expected behavior

In the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#precision), the video tells us that the weights remain FP32 and the data is cast to the selected precision. When training with FP16 data on IPU, the cast should occur before copying the data to device. The screenshots below show that the data is FP32 during the copy when `precision=16`.

### Environment

* Packages:
        - numpy:             1.23.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cpu
        - pytorch-lightning: 1.6.5
        - tqdm:              4.64.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #124-Ubuntu SMP Thu Apr 14 19:46:19 UTC 2022

### Additional context

FP16 - left, FP32 - right

In this image you can see that the total amount of data received by the IPU for this batch is 301184B in both cases. If we divide by the size of the batch, it can be seen that the datatype in both copies is FP32. `301184 / (32*3*28*28) = 4B = 32b`:
![image](https://user-images.githubusercontent.com/19981378/180759494-91f71928-ead7-48db-8aea-4cbe8db3aded.png)

cc @carmocca @justusschock @awaelchli @akihironitta @rohitgr7

**Comment by rohitgr7** [comment]
a simple solution I can think of is to override
https://github.com/Lightning-AI/lightning/blob/227871982dd44ef9e6e060e7bd16043b99d8087e/src/pytorch_lightning/strategies/strategy.py#L216

and cast the data if precision is set before moving it to the device.

also noticed similar behavior with deepspeed too.

just wondering if there are any implications of doing that. Can't think of any.

**Comment by HMellor** [comment]
We should probably also remove the `tensor.to(device)` as that doesn't do anything when using IPUs. How would we go about overriding this function so that it does nothing but cast the data to the precision specified in the Trainer?

**Comment by carmocca** [comment]
You can override the method that Rohit linked by Rohit:

```py
class MyWorkaroudnIPUStrategy(IPUStrategy):
    def batch_to_device(...):
        ...
```

**Comment by awaelchli** [comment]
Thanks for the PR @HMellor 
I will label this as a bug if I understand correctly.
